File: binding/include/libqundostack.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qundocommand.html
pub const qundocommand = struct {

    /// New constructs a new QUndoCommand object.
    pub fn New() QtC.QUndoCommand {
        return qtc.QUndoCommand_new();
    }


    /// New2 constructs a new QUndoCommand object.
    pub fn New2(text: []const u8) QtC.QUndoCommand {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
return qtc.QUndoCommand_new2(text_str);
    }


    /// New3 constructs a new QUndoCommand object.
    pub fn New3(parent: ?*anyopaque) QtC.QUndoCommand {
        return qtc.QUndoCommand_new3(@ptrCast(parent));
    }


    /// New4 constructs a new QUndoCommand object.
    pub fn New4(text: []const u8, parent: ?*anyopaque) QtC.QUndoCommand {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
return qtc.QUndoCommand_new4(text_str, @ptrCast(parent));
    }


    pub fn Undo(self: ?*anyopaque, ) void {
        qtc.QUndoCommand_Undo(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnUndo(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QUndoCommand_OnUndo(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseUndo(self: ?*anyopaque, ) void {
        qtc.QUndoCommand_QBaseUndo(@ptrCast(self));
    }

    pub fn Redo(self: ?*anyopaque, ) void {
        qtc.QUndoCommand_Redo(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnRedo(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QUndoCommand_OnRedo(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseRedo(self: ?*anyopaque, ) void {
        qtc.QUndoCommand_QBaseRedo(@ptrCast(self));
    }

    pub fn Text(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUndoCommand_Text(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qundocommand.Text: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ActionText(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUndoCommand_ActionText(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qundocommand.ActionText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetText(self: ?*anyopaque, text: []const u8) void {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
qtc.QUndoCommand_SetText(@ptrCast(self), text_str);
    }

    pub fn IsObsolete(self: ?*anyopaque, ) bool {
        return qtc.QUndoCommand_IsObsolete(@ptrCast(self));
    }

    pub fn SetObsolete(self: ?*anyopaque, obsolete: bool) void {
        qtc.QUndoCommand_SetObsolete(@ptrCast(self), obsolete);
    }

    pub fn Id(self: ?*anyopaque, ) i32 {
        return qtc.QUndoCommand_Id(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnId(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QUndoCommand_OnId(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseId(self: ?*anyopaque, ) i32 {
        return qtc.QUndoCommand_QBaseId(@ptrCast(self));
    }

    pub fn MergeWith(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QUndoCommand_MergeWith(@ptrCast(self), @ptrCast(other));
    }

    /// Allows for overriding the related default method
    pub fn OnMergeWith(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QUndoCommand_OnMergeWith(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMergeWith(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QUndoCommand_QBaseMergeWith(@ptrCast(self), @ptrCast(other));
    }

    pub fn ChildCount(self: ?*anyopaque, ) i32 {
        return qtc.QUndoCommand_ChildCount(@ptrCast(self));
    }

    pub fn Child(self: ?*anyopaque, index: i32) QtC.QUndoCommand {
        return qtc.QUndoCommand_Child(@ptrCast(self), index);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QUndoCommand_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qundostack.html
pub const qundostack = struct {

    /// New constructs a new QUndoStack object.
    pub fn New() QtC.QUndoStack {
        return qtc.QUndoStack_new();
    }


    /// New2 constructs a new QUndoStack object.
    pub fn New2(parent: ?*anyopaque) QtC.QUndoStack {
        return qtc.QUndoStack_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUndoStack_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qundostack.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QUndoStack_Clear(@ptrCast(self));
    }

    pub fn Push(self: ?*anyopaque, cmd: ?*anyopaque) void {
        qtc.QUndoStack_Push(@ptrCast(self), @ptrCast(cmd));
    }

    pub fn CanUndo(self: ?*anyopaque, ) bool {
        return qtc.QUndoStack_CanUndo(@ptrCast(self));
    }

    pub fn CanRedo(self: ?*anyopaque, ) bool {
        return qtc.QUndoStack_CanRedo(@ptrCast(self));
    }

    pub fn UndoText(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUndoStack_UndoText(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qundostack.UndoText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn RedoText(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUndoStack_RedoText(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qundostack.RedoText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Count(self: ?*anyopaque, ) i32 {
        return qtc.QUndoStack_Count(@ptrCast(self));
    }

    pub fn Index(self: ?*anyopaque, ) i32 {
        return qtc.QUndoStack_Index(@ptrCast(self));
    }

    pub fn Text(self: ?*anyopaque, idx: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUndoStack_Text(@ptrCast(self), idx);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qundostack.Text: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn CreateUndoAction(self: ?*anyopaque, parent: ?*anyopaque) QtC.QAction {
        return qtc.QUndoStack_CreateUndoAction(@ptrCast(self), @ptrCast(parent));
    }

    pub fn CreateRedoAction(self: ?*anyopaque, parent: ?*anyopaque) QtC.QAction {
        return qtc.QUndoStack_CreateRedoAction(@ptrCast(self), @ptrCast(parent));
    }

    pub fn IsActive(self: ?*anyopaque, ) bool {
        return qtc.QUndoStack_IsActive(@ptrCast(self));
    }

    pub fn IsClean(self: ?*anyopaque, ) bool {
        return qtc.QUndoStack_IsClean(@ptrCast(self));
    }

    pub fn CleanIndex(self: ?*anyopaque, ) i32 {
        return qtc.QUndoStack_CleanIndex(@ptrCast(self));
    }

    pub fn BeginMacro(self: ?*anyopaque, text: []const u8) void {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
qtc.QUndoStack_BeginMacro(@ptrCast(self), text_str);
    }

    pub fn EndMacro(self: ?*anyopaque, ) void {
        qtc.QUndoStack_EndMacro(@ptrCast(self));
    }

    pub fn SetUndoLimit(self: ?*anyopaque, limit: i32) void {
        qtc.QUndoStack_SetUndoLimit(@ptrCast(self), limit);
    }

    pub fn UndoLimit(self: ?*anyopaque, ) i32 {
        return qtc.QUndoStack_UndoLimit(@ptrCast(self));
    }

    pub fn Command(self: ?*anyopaque, index: i32) QtC.QUndoCommand {
        return qtc.QUndoStack_Command(@ptrCast(self), index);
    }

    pub fn SetClean(self: ?*anyopaque, ) void {
        qtc.QUndoStack_SetClean(@ptrCast(self));
    }

    pub fn ResetClean(self: ?*anyopaque, ) void {
        qtc.QUndoStack_ResetClean(@ptrCast(self));
    }

    pub fn SetIndex(self: ?*anyopaque, idx: i32) void {
        qtc.QUndoStack_SetIndex(@ptrCast(self), idx);
    }

    pub fn Undo(self: ?*anyopaque, ) void {
        qtc.QUndoStack_Undo(@ptrCast(self));
    }

    pub fn Redo(self: ?*anyopaque, ) void {
        qtc.QUndoStack_Redo(@ptrCast(self));
    }

    pub fn SetActive(self: ?*anyopaque, ) void {
        qtc.QUndoStack_SetActive(@ptrCast(self));
    }

    pub fn IndexChanged(self: ?*anyopaque, idx: i32) void {
        qtc.QUndoStack_IndexChanged(@ptrCast(self), idx);
    }

    pub fn OnIndexChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QUndoStack_Connect_IndexChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn CleanChanged(self: ?*anyopaque, clean: bool) void {
        qtc.QUndoStack_CleanChanged(@ptrCast(self), clean);
    }

    pub fn OnCleanChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QUndoStack_Connect_CleanChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn CanUndoChanged(self: ?*anyopaque, canUndo: bool) void {
        qtc.QUndoStack_CanUndoChanged(@ptrCast(self), canUndo);
    }

    pub fn OnCanUndoChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QUndoStack_Connect_CanUndoChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn CanRedoChanged(self: ?*anyopaque, canRedo: bool) void {
        qtc.QUndoStack_CanRedoChanged(@ptrCast(self), canRedo);
    }

    pub fn OnCanRedoChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QUndoStack_Connect_CanRedoChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn UndoTextChanged(self: ?*anyopaque, undoText: []const u8) void {
        const undoText_str = qtc.libqt_string{
    .len = undoText.len,
    .data = undoText.ptr,
};
qtc.QUndoStack_UndoTextChanged(@ptrCast(self), undoText_str);
    }

    pub fn OnUndoTextChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) void) void {
        qtc.QUndoStack_Connect_UndoTextChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn RedoTextChanged(self: ?*anyopaque, redoText: []const u8) void {
        const redoText_str = qtc.libqt_string{
    .len = redoText.len,
    .data = redoText.ptr,
};
qtc.QUndoStack_RedoTextChanged(@ptrCast(self), redoText_str);
    }

    pub fn OnRedoTextChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) void) void {
        qtc.QUndoStack_Connect_RedoTextChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUndoStack_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qundostack.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUndoStack_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qundostack.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn CreateUndoAction2(self: ?*anyopaque, parent: ?*anyopaque, prefix: []const u8) QtC.QAction {
        const prefix_str = qtc.libqt_string{
    .len = prefix.len,
    .data = prefix.ptr,
};
return qtc.QUndoStack_CreateUndoAction2(@ptrCast(self), @ptrCast(parent), prefix_str);
    }

    pub fn CreateRedoAction2(self: ?*anyopaque, parent: ?*anyopaque, prefix: []const u8) QtC.QAction {
        const prefix_str = qtc.libqt_string{
    .len = prefix.len,
    .data = prefix.ptr,
};
return qtc.QUndoStack_CreateRedoAction2(@ptrCast(self), @ptrCast(parent), prefix_str);
    }

    pub fn SetActive1(self: ?*anyopaque, active: bool) void {
        qtc.QUndoStack_SetActive1(@ptrCast(self), active);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QUndoStack_Delete(@ptrCast(self));
    }
};

File: binding/src/libqundostack.cpp
#include <QAction>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QUndoCommand>
#include <QUndoStack>
#include <qundostack.h>
#include "libqundostack.h"
#include "libqundostack.hxx"

QUndoCommand* QUndoCommand_new() {
	 return new VirtualQUndoCommand();
}

QUndoCommand* QUndoCommand_new2(const libqt_string text) {
	 return new VirtualQUndoCommand(QString::fromUtf8(text.data, text.len));
}

QUndoCommand* QUndoCommand_new3(QUndoCommand* parent) {
	 return new VirtualQUndoCommand(parent);
}

QUndoCommand* QUndoCommand_new4(const libqt_string text, QUndoCommand* parent) {
	 return new VirtualQUndoCommand(QString::fromUtf8(text.data, text.len), parent);
}

void QUndoCommand_Undo(QUndoCommand* self) {
	self->undo();
}

void QUndoCommand_Redo(QUndoCommand* self) {
	self->redo();
}

libqt_string QUndoCommand_Text(const QUndoCommand* self) {
	QString _ret = self->text();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QUndoCommand_ActionText(const QUndoCommand* self) {
	QString _ret = self->actionText();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QUndoCommand_SetText(QUndoCommand* self, const libqt_string text) {
	self->setText(QString::fromUtf8(text.data, text.len));
}

bool QUndoCommand_IsObsolete(const QUndoCommand* self) {
	return self->isObsolete();
}

void QUndoCommand_SetObsolete(QUndoCommand* self, bool obsolete) {
	self->setObsolete(obsolete);
}

int QUndoCommand_Id(const QUndoCommand* self) {
	return self->id();
}

bool QUndoCommand_MergeWith(QUndoCommand* self, const QUndoCommand* other) {
	return self->mergeWith(other);
}

int QUndoCommand_ChildCount(const QUndoCommand* self) {
	return self->childCount();
}

const QUndoCommand* QUndoCommand_Child(const QUndoCommand* self, int index) {
	return self->child(index);
}

// Base class handler implementation
void QUndoCommand_QBaseUndo(QUndoCommand* self) {
	auto* vqundocommand = dynamic_cast<VirtualQUndoCommand*>(self);
	if (vqundocommand && vqundocommand->isVirtualQUndoCommand) {
vqundocommand->setQUndoCommand_Undo_IsBase(true);
	vqundocommand->undo();
}
}

// Auxiliary method to allow providing re-implementation
void QUndoCommand_OnUndo(QUndoCommand* self, intptr_t slot) {
	auto* vqundocommand = dynamic_cast<VirtualQUndoCommand*>(self);
	if (vqundocommand && vqundocommand->isVirtualQUndoCommand) {
vqundocommand->setQUndoCommand_Undo_Callback(reinterpret_cast<VirtualQUndoCommand::QUndoCommand_Undo_Callback>(slot));
}
}

// Base class handler implementation
void QUndoCommand_QBaseRedo(QUndoCommand* self) {
	auto* vqundocommand = dynamic_cast<VirtualQUndoCommand*>(self);
	if (vqundocommand && vqundocommand->isVirtualQUndoCommand) {
vqundocommand->setQUndoCommand_Redo_IsBase(true);
	vqundocommand->redo();
}
}

// Auxiliary method to allow providing re-implementation
void QUndoCommand_OnRedo(QUndoCommand* self, intptr_t slot) {
	auto* vqundocommand = dynamic_cast<VirtualQUndoCommand*>(self);
	if (vqundocommand && vqundocommand->isVirtualQUndoCommand) {
vqundocommand->setQUndoCommand_Redo_Callback(reinterpret_cast<VirtualQUndoCommand::QUndoCommand_Redo_Callback>(slot));
}
}

// Base class handler implementation
int QUndoCommand_QBaseId(const QUndoCommand* self) {
	auto* vqundocommand = dynamic_cast<const VirtualQUndoCommand*>(self);
	if (vqundocommand && vqundocommand->isVirtualQUndoCommand) {
vqundocommand->setQUndoCommand_Id_IsBase(true);
	return vqundocommand->id();
}
}

// Auxiliary method to allow providing re-implementation
void QUndoCommand_OnId(const QUndoCommand* self, intptr_t slot) {
	auto* vqundocommand = dynamic_cast<const VirtualQUndoCommand*>(self);
	if (vqundocommand && vqundocommand->isVirtualQUndoCommand) {
vqundocommand->setQUndoCommand_Id_Callback(reinterpret_cast<VirtualQUndoCommand::QUndoCommand_Id_Callback>(slot));
}
}

// Base class handler implementation
bool QUndoCommand_QBaseMergeWith(QUndoCommand* self, const QUndoCommand* other) {
	auto* vqundocommand = dynamic_cast<VirtualQUndoCommand*>(self);
	if (vqundocommand && vqundocommand->isVirtualQUndoCommand) {
vqundocommand->setQUndoCommand_MergeWith_IsBase(true);
	return vqundocommand->mergeWith(other);
}
}

// Auxiliary method to allow providing re-implementation
void QUndoCommand_OnMergeWith(QUndoCommand* self, intptr_t slot) {
	auto* vqundocommand = dynamic_cast<VirtualQUndoCommand*>(self);
	if (vqundocommand && vqundocommand->isVirtualQUndoCommand) {
vqundocommand->setQUndoCommand_MergeWith_Callback(reinterpret_cast<VirtualQUndoCommand::QUndoCommand_MergeWith_Callback>(slot));
}
}

void QUndoCommand_Delete(QUndoCommand* self) {
    delete self;
}

QUndoStack* QUndoStack_new() {
	 return new QUndoStack();
}

QUndoStack* QUndoStack_new2(QObject* parent) {
	 return new QUndoStack(parent);
}

libqt_string QUndoStack_Tr(const char* s) {
	QString _ret = QUndoStack::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QUndoStack_Clear(QUndoStack* self) {
	self->clear();
}

void QUndoStack_Push(QUndoStack* self, QUndoCommand* cmd) {
	self->push(cmd);
}

bool QUndoStack_CanUndo(const QUndoStack* self) {
	return self->canUndo();
}

bool QUndoStack_CanRedo(const QUndoStack* self) {
	return self->canRedo();
}

libqt_string QUndoStack_UndoText(const QUndoStack* self) {
	QString _ret = self->undoText();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QUndoStack_RedoText(const QUndoStack* self) {
	QString _ret = self->redoText();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QUndoStack_Count(const QUndoStack* self) {
	return self->count();
}

int QUndoStack_Index(const QUndoStack* self) {
	return self->index();
}

libqt_string QUndoStack_Text(const QUndoStack* self, int idx) {
	QString _ret = self->text(idx);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QAction* QUndoStack_CreateUndoAction(const QUndoStack* self, QObject* parent) {
	return self->createUndoAction(parent);
}

QAction* QUndoStack_CreateRedoAction(const QUndoStack* self, QObject* parent) {
	return self->createRedoAction(parent);
}

bool QUndoStack_IsActive(const QUndoStack* self) {
	return self->isActive();
}

bool QUndoStack_IsClean(const QUndoStack* self) {
	return self->isClean();
}

int QUndoStack_CleanIndex(const QUndoStack* self) {
	return self->cleanIndex();
}

void QUndoStack_BeginMacro(QUndoStack* self, const libqt_string text) {
	self->beginMacro(QString::fromUtf8(text.data, text.len));
}

void QUndoStack_EndMacro(QUndoStack* self) {
	self->endMacro();
}

void QUndoStack_SetUndoLimit(QUndoStack* self, int limit) {
	self->setUndoLimit(limit);
}

int QUndoStack_UndoLimit(const QUndoStack* self) {
	return self->undoLimit();
}

const QUndoCommand* QUndoStack_Command(const QUndoStack* self, int index) {
	return self->command(index);
}

void QUndoStack_SetClean(QUndoStack* self) {
	self->setClean();
}

void QUndoStack_ResetClean(QUndoStack* self) {
	self->resetClean();
}

void QUndoStack_SetIndex(QUndoStack* self, int idx) {
	self->setIndex(idx);
}

void QUndoStack_Undo(QUndoStack* self) {
	self->undo();
}

void QUndoStack_Redo(QUndoStack* self) {
	self->redo();
}

void QUndoStack_SetActive(QUndoStack* self) {
	self->setActive();
}

void QUndoStack_IndexChanged(QUndoStack* self, int idx) {
	self->indexChanged(idx);
}

void QUndoStack_Connect_IndexChanged(QUndoStack* self, intptr_t slot) {
    void (*slotFunc)(QUndoStack*, int) = reinterpret_cast<void (*)(QUndoStack*, int)>(slot);
    QUndoStack::connect(self, &QUndoStack::indexChanged, [self, slotFunc](int idx) {
	slotFunc(self, idx);
    });
}

void QUndoStack_CleanChanged(QUndoStack* self, bool clean) {
	self->cleanChanged(clean);
}

void QUndoStack_Connect_CleanChanged(QUndoStack* self, intptr_t slot) {
    void (*slotFunc)(QUndoStack*, bool) = reinterpret_cast<void (*)(QUndoStack*, bool)>(slot);
    QUndoStack::connect(self, &QUndoStack::cleanChanged, [self, slotFunc](bool clean) {
	slotFunc(self, clean);
    });
}

void QUndoStack_CanUndoChanged(QUndoStack* self, bool canUndo) {
	self->canUndoChanged(canUndo);
}

void QUndoStack_Connect_CanUndoChanged(QUndoStack* self, intptr_t slot) {
    void (*slotFunc)(QUndoStack*, bool) = reinterpret_cast<void (*)(QUndoStack*, bool)>(slot);
    QUndoStack::connect(self, &QUndoStack::canUndoChanged, [self, slotFunc](bool canUndo) {
	slotFunc(self, canUndo);
    });
}

void QUndoStack_CanRedoChanged(QUndoStack* self, bool canRedo) {
	self->canRedoChanged(canRedo);
}

void QUndoStack_Connect_CanRedoChanged(QUndoStack* self, intptr_t slot) {
    void (*slotFunc)(QUndoStack*, bool) = reinterpret_cast<void (*)(QUndoStack*, bool)>(slot);
    QUndoStack::connect(self, &QUndoStack::canRedoChanged, [self, slotFunc](bool canRedo) {
	slotFunc(self, canRedo);
    });
}

void QUndoStack_UndoTextChanged(QUndoStack* self, const libqt_string undoText) {
	self->undoTextChanged(QString::fromUtf8(undoText.data, undoText.len));
}

void QUndoStack_Connect_UndoTextChanged(QUndoStack* self, intptr_t slot) {
    void (*slotFunc)(QUndoStack*, const char*) = reinterpret_cast<void (*)(QUndoStack*, const char*)>(slot);
    QUndoStack::connect(self, &QUndoStack::undoTextChanged, [self, slotFunc](const QString& undoText) {
	slotFunc(self, undoText);
    });
}

void QUndoStack_RedoTextChanged(QUndoStack* self, const libqt_string redoText) {
	self->redoTextChanged(QString::fromUtf8(redoText.data, redoText.len));
}

void QUndoStack_Connect_RedoTextChanged(QUndoStack* self, intptr_t slot) {
    void (*slotFunc)(QUndoStack*, const char*) = reinterpret_cast<void (*)(QUndoStack*, const char*)>(slot);
    QUndoStack::connect(self, &QUndoStack::redoTextChanged, [self, slotFunc](const QString& redoText) {
	slotFunc(self, redoText);
    });
}

libqt_string QUndoStack_Tr2(const char* s, const char* c) {
	QString _ret = QUndoStack::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QUndoStack_Tr3(const char* s, const char* c, int n) {
	QString _ret = QUndoStack::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QAction* QUndoStack_CreateUndoAction2(const QUndoStack* self, QObject* parent, const libqt_string prefix) {
	return self->createUndoAction(parent, QString::fromUtf8(prefix.data, prefix.len));
}

QAction* QUndoStack_CreateRedoAction2(const QUndoStack* self, QObject* parent, const libqt_string prefix) {
	return self->createRedoAction(parent, QString::fromUtf8(prefix.data, prefix.len));
}

void QUndoStack_SetActive1(QUndoStack* self, bool active) {
	self->setActive(active);
}

void QUndoStack_Delete(QUndoStack* self) {
    delete self;
}

